osu.Game/Online/Multiplayer/OnlineMultiplayerClient.cs

```diff
@@ -71,14 +73,23 @@ private void load(IAPIProvider api)
             }
         }
 
-        protected override Task<MultiplayerRoom> JoinRoom(long roomId, string? password = null)
+        protected override async Task<MultiplayerRoom> JoinRoom(long roomId, string? password = null)
         {
             if (!IsConnected.Value)
-                return Task.FromCanceled<MultiplayerRoom>(new CancellationToken(true));
+                throw new OperationCanceledException();
```

Comment:
I'm not completely sure this throw will behave the same, behaviourally. I get why this was changed (can't return a `Task` from within an `async` method), but I believe this change may cause the task to transition into a `Faulted` state rather than `Canceled`, judging by [MS docs](https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/task-cancellation):

> When a task instance observes an [OperationCanceledException](https://docs.microsoft.com/en-us/dotnet/api/system.operationcanceledexception) thrown by user code, it compares the exception's token to its associated token (the one that was passed to the API that created the Task). If they are the same and the token's [IsCancellationRequested](https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtoken.iscancellationrequested) property returns true, the task interprets this as acknowledging cancellation and transitions to the Canceled state. If you do not use a [Wait](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.wait) or [WaitAll](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.waitall) method to wait for the task, then the task just sets its status to [Canceled](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskstatus#system-threading-tasks-taskstatus-canceled).

Not sure if it's a big deal - the cancellation may end up being logged as error, but (a) I'm not sure if it matters, and (b) not sure how to fix it either.

Reply:
Is there a better alternative that more closely matches the previous behaviour?

Reply:
I don't think so, which is why I'm not going to make a blocker out of this or anything.